### PR TITLE
Append c10 libs to TorchConfig.cmake

### LINKED ISSUE #18416: TorchConfig does not add c10 to TORCH_LIBRARIES
## 🐛 Bug

When using find_package(Torch CONFIG) and attempting to use the resulting TORCH_LIBRARIES, I find that c10 is not linked in.

The result is my programs will complain about undefined references to all c10 symbols (such as c10::Error, which is thrown from torch::jit::load in libtorch).

## To Reproduce

1. Create sample app that loads a model in libtorch
2. Have the CMakeLists find_package(Torch CONFIG) and link the app to TORCH_LIBRARIES
3. When compiling, notice undefined references

Code:
```C++
#include <torch/script.h>
int main() {
    std::shared_ptr<torch::jit::script::Module> model = torch::jit::load(model_path);
}
```

```CMake
add_executable(test main.cpp)
find_package(Torch CONFIG REQUIRED)
target_link_libraries(test ${TORCH_LIBRARIES})
```

## Expected behavior

Links to what is required

## Environment

 - PyTorch Version (e.g., 1.0): master / nightly
 - OS (e.g., Linux): Windows
 - How you installed PyTorch (`conda`, `pip`, source): Tried from source and libtorch from website
 - Build command you used (if compiling from source):
 - Python version: 3.7.2
 - CUDA/cuDNN version: 10.0
 - GPU models and configuration: GTX 1060

## Additional context

Workaround of:
```CMake
set(TORCH_LIBRARIES ${TORCH_LIBRARIES} "${TORCH_ROOT}/lib/c10.lib" "${TORCH_ROOT}/lib/c10_cuda.lib")
```
makes everything work.

### PR DESCRIPTION
**Pull request by peterjc123**
Fixes #18416.

### CODE REVIEW

**Comment by xsacha** [comment]
Is this like the Caffe2 libs where it would not need to be added to linker list if built statically?
Not sure if it needs wrapping with `if (@BUILD_SHARED_LIBS@)`.

**Comment by peterjc123** [comment]
> Is this like the Caffe2 libs where it would not need to be added to linker list if built statically?
> Not sure if it needs wrapping with `if (@BUILD_SHARED_LIBS@)`.
> 
> On Linux I build statically for example and did not need this added. I discovered this issue on Windows as static was broken and the nightlies are shared.

Actually, it should be `if (NOT @BUILD_SHARED_LIBS@).`, but it won't be a problem since, in shared builds, the symbols are all contained in `torch.dll` so adding any other libraries will have no effect. The linker will remove useless libraries specified by the user automatically.

**Comment by ezyang** [comment]
@peterjc123 This seems fine, but is there any chance we could get a test for this? This seems like the kind of thing we'd be likely to break upon a cmake refactor without a test for it.

**Review by ezyang — APPROVED** [review]

File: cmake/TorchConfig.cmake.in [hunk 1]
```diff
@@ -42,6 +42,13 @@ find_library(TORCH_LIBRARY torch PATHS "${TORCH_INSTALL_PREFIX}/lib")
 add_library(torch UNKNOWN IMPORTED)
 set(TORCH_LIBRARIES torch ${Caffe2_MAIN_LIBS})
 
+find_library(C10_LIBRARY c10 PATHS "${TORCH_INSTALL_PREFIX}/lib")
+list(APPEND TORCH_LIBRARIES ${C10_LIBRARY})
+if (@USE_CUDA@)
+  find_library(C10_CUDA_LIBRARY c10_cuda PATHS "${TORCH_INSTALL_PREFIX}/lib")
+  list(APPEND TORCH_LIBRARIES ${C10_CUDA_LIBRARY})
```
Comment (ezyang) [inline cmake/TorchConfig.cmake.in]: Actually, are you sure this shouldn't be `TORCH_CUDA_LIBRARIES`?

**Review by ezyang — CHANGES_REQUESTED** [review]
one question

**Comment by peterjc123** [comment]
@ezyang About the test, I'll do it in another PR. Is it a good idea to add the cpp-extensions tests to `caffe2-py2-cuda9.0-cudnn7-windows-build` and change it to use `build_libtorch.py` instead of `build_windows.bat`? cc @orionr

**Comment by orionr** [comment]
Moving to `build_libtorch.py` would be a win, IMO. cc @kostmo and @pjh5

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**